hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/utils/DescriptorUtils.java

```diff
@@ -57,11 +59,32 @@ public static boolean isViewFunction(final Bytes input) {
         };
     }
 
-    public static RedirectTarget getRedirectTarget(final Bytes input) {
-        final var tokenAddress = input.slice(4, 20);
+    public static RedirectTarget getRedirectTarget(Bytes input) {
+        final var finalInput = getFinalInput(input);
+        final var tokenAddress = finalInput.slice(4, 20);
         final var tokenId = tokenIdFromEvmAddress(tokenAddress.toArrayUnsafe());
-        final var nestedInput = input.slice(24);
-        return new RedirectTarget(nestedInput.getInt(0), tokenId);
+        final var nestedInput = finalInput.slice(24);
+        return new RedirectTarget(nestedInput.getInt(0), tokenId, finalInput);
+    }
+
+    private static Bytes getFinalInput(final Bytes input) {
+        try {
+            // try decoding the input to see if redirectForToken() was called explicitly, using
+            // normal encoding
+            // if so, massage it to our expected "packed" redirect input
+            final var tuple =
+                    EvmDecodingFacade.decodeFunctionCall(
+                            input, REDIRECT_FOR_TOKEN_SELECTOR, REDIRECT_FOR_TOKEN_DECODER);
+            final var massagedInput =
+                    com.google.common.primitives.Bytes.concat(
```

Comment:
We can use `org.apache.tuweni.bytes.Bytes.concatenate` instead?

Reply:
Yup, changed now